src/python/pants/backend/cc/util_rules/toolchain.py

```diff
@@ -0,0 +1,212 @@
+        # Or just ensure all defines have -D right now?
+        if self.compile_defines:
+            sanitized_defines = [define.lstrip("-D") for define in self.compile_defines]
+            object.__setattr__(self, "compile_defines", tuple(sanitized_defines))
+
+    @property
+    def compile_command(self) -> tuple[str, ...]:
+        """The command to compile a C/C++ source file."""
+        command = [self.compiler, *self.compile_defines, *self.compile_flags]
+        return tuple(filter(None, command))
+
+    @property
+    def link_command(self) -> tuple[str, ...]:
+        """The command to link a C/C++ binary."""
+        command = [self.compiler, *self.link_flags]
+        return tuple(filter(None, command))
+
+
+@rule_helper
+async def _executable_path(binary_names: Iterable[str], search_paths: Iterable[str]) -> str:
+    """Find the path to an executable by checking whether the executable supports a version
+    option."""
+    for name in binary_names:
+        binary_paths = await Get(
+            BinaryPaths,
+            BinaryPathRequest(
+                binary_name=name,
+                search_path=search_paths,
+                test=BinaryPathTest(args=["-v"]),
+            ),
+        )
+
+        if not binary_paths or not binary_paths.first_path:
+            continue
+        return binary_paths.first_path.path
+
+    raise BinaryNotFoundError(f"Could not find any of '{binary_names}' in any of {search_paths}.")
+
+
+@rule_helper
+async def _setup_downloadable_toolchain(
+    request: CCToolchainRequest,
+    subsystem: ExternalCCSubsystem,
+    platform: Platform,
+) -> CCToolchain:
+    """Set up a toolchain from a downloadable archive."""
+
+    download_file_request = subsystem.get_request(platform).download_file_request
+    maybe_archive_digest = await Get(Digest, DownloadFile, download_file_request)
+    extracted_archive = await Get(ExtractedArchive, Digest, maybe_archive_digest)
+
+    # Populate the toolchain for C or C++ accordingly
+    if request.language == CCLanguage.CPP:
+        return CCToolchain(
+            compiler=subsystem.cpp_executable,
+            compile_flags=tuple(subsystem.cpp_compile_options),
+            compile_defines=tuple(subsystem.cpp_defines),
+            digest=extracted_archive.digest,
+        )
+
+    return CCToolchain(
+        compiler=subsystem.c_executable,
+        compile_flags=tuple(subsystem.c_compile_options),
+        compile_defines=tuple(subsystem.c_defines),
+        digest=extracted_archive.digest,
+    )
+
+
+@rule_helper
+async def _setup_system_toolchain(
+    request: CCToolchainRequest, subsystem: CCSubsystem
+) -> CCToolchain:
+    """Set up a toolchain from the user's host system."""
+
+    # Sanitize the search paths in case the "<PATH>" is specified
+    raw_search_paths = list(subsystem.search_paths)
+    if "<PATH>" in raw_search_paths:
+        i = raw_search_paths.index("<PATH>")
+        env = await Get(Environment, EnvironmentRequest(["PATH"]))
```

Comment:
The environment rules took over the naming here. These are now called `EnvironmentVars` and `EnvironmentVarsRequest` respectively.

Reply:
👍🏽  Thanks! I didn't know that. I saw all the Environment work, but didn't see what the end result was.